Thumbnail support in Zulip

adnrs96 opened the pull request:
This PR is going to track our progress upon bringing thumbnail support to Zulip. This will benefit mobile devices in terms of performance. PR's related to this are #4973  #2590 and issue related is #432.

smarx commented:
*Automated message from Dropbox CLA bot*

@adnrs96, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on Vagrantfile:
```diff
@@ -49,6 +49,8 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   end
 
   config.vm.network "forwarded_port", guest: 9991, host: host_port, host_ip: host_ip_addr
+  config.vm.network "forwarded_port", guest: 9995, host: 9995, host_ip: host_ip_addr
```
We should have thumbor be proxied by `run-dev.py`, similar to how tornado is, in which case we don't need this line.

adnrs96 replied:
yup, I missed out one thing regarding nginx proxying which is clear now reading this comment so will fix that.

timabbott commented on puppet/zulip/files/supervisor/conf.d/thumbor.conf:
```diff
@@ -0,0 +1,14 @@
+stdout_logfile_backups=10
+stderr_logfile=/var/log/zulip/thumbor.stderr.log
+stderr_logfile_maxbytes=1MB
+stderr_logfile_backups=10
```
I would follow the convention of our other supervisord files for stdout/stderr, namely, putting both stderr and stdout in the same file in `/var/log/zulip/thumbor.log` via `redirect_stderr`, and having the log rotation be 20MB x 3.

timabbott commented on puppet/zulip/manifests/thumbor.pp:
```diff
@@ -0,0 +1,40 @@
+# Class: zulip::thumbor
+#
+#
+class zulip::thumbor {
```
This module should `include zulip::supervisor` and `include zulip::nginx`, since it in practice depends on them.

timabbott commented on requirements/thumbor.txt:
```diff
@@ -0,0 +1,24 @@
+thumbor==6.1.5
+tornado==4.4.2
+tornado-botocore==1.0.2
+typing
```
I think it'd be nice to use have a `thumbor_locked.txt` that is generated from a basic `thumbor.txt` dependency, so that we don't have to maintain this list of dependency versions by hand.

See `tools/update-locked-requirements` for how we do this for other things.

adnrs96 replied:
I have included a thumbor_lock.txt but haven't updated `tools/update-locked-requirements` to auto generate/update it basically because I didn't knew if that's expected and also because I couldn't account for how the py2_common_lock.txt is autogenerated if it at all.

timabbott commented on scripts/lib/install:
```diff
@@ -137,6 +138,7 @@ if [ "$has_appserver" = 0 ]; then
     ln -nsf "$deploy_path" /home/zulip/deployments/next
     ln -nsf "$deploy_path" /home/zulip/deployments/current
     ln -nsf /etc/zulip/settings.py "$deploy_path"/zproject/prod_settings.py
+    ln -nsf "$deploy_path"/zulip-thumbor-venv/bin/thumbor /usr/local/bin/thumbor
```
I'm not sure this symlink makes sense to me; why not just have things using thumbor write `/home/zulip/deployments/current/zulip-thumbor-venv/bin/thumbor`?

timabbott commented on scripts/setup/generate_secrets.py:
```diff
@@ -33,6 +33,7 @@
     'local_database_password',
     'rabbitmq_password',
     'shared_secret',
+    'thumbor_sign_key',
```
Given the secret names in the rest of Zulip, I think probably `thumbor_key` would be what we'd usually call such a thing; does it make sense to rename this?

timabbott commented on tools/lib/provision.py:
```diff
@@ -126,6 +126,7 @@
     "gettext",              # Used by makemessages i18n
     "curl",                 # Used for fetching PhantomJS as wget occasionally fails on redirects
     "netcat",               # Used for flushing memcached
+    "libcurl4-openssl-dev",  # Thumbor Required Library
```
Why is this list of libraries so much shorter than the one in `puppet`?  Makes me wonder if we're missing some that should be here.

timabbott commented on tools/run-dev.py:
```diff
@@ -159,9 +159,13 @@
 os.chdir(os.path.join(os.path.dirname(__file__), '..'))
 
+# Thumbor server address
+os.environ['THUMBOR_HOST'] = '127.0.0.1:{0}'.format(thumbor_port)
```
Does this environment variable do anything?  I don't see where it does, so it might be dead code.

timabbott commented on tools/setup/setup_venvs.py:
```diff
@@ -15,11 +15,14 @@
 
 def main(is_travis=False):
     # type: (bool) -> None
+    setup_virtualenv("/srv/thumbor-venv", THUMBOR_REQS_FILE, patch_activate_script=True)
```
Since we're planning to move to Python 3 only, let's start with Thumbor as Python 3 only.

adnrs96 replied:
For the sake of parity in discussion here, I would like to post an update on this here as well.
So basically problem with addressing this is that thumbor doesn't support python 3 yet as per https://github.com/thumbor/thumbor/issues/738

So for now it was [decided](https://chat.zulip.org/#narrow/stream/backend/subject/thumbor/near/264744) that we just keep thumbor in different venv and make sure its interfacing with zulip is py3 compatible and let thumbor live in py2.

timabbott commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -144,7 +144,12 @@ def add_a(root, url, link, title=None, desc=None,
-    img.set("src", url)
+    thumbnail_path = '/thumbnail'
+    if not url.startswith('/'):
+        thumbnail_path += '/'
```
It might be cleaner if this used `urljoin` instead?

adnrs96 replied:
I believe this should be the right way. Using `urljoin` doesn't give an expected behaviour here always. What I am doing here is that any incoming link that will have a preview image will have a img src set to call the `/thumbnail` endpoint. 
Now using `urljoin` creates problems here since it joins relatively and doesn't full fill expectations here.

timabbott replied:
OK; we should make sure that we have Bugdown tests covering the different cases in this function, to protect us from breaking things in a future refactoring.

adnrs96 replied:
Sure.

showell commented on zerver/lib/thumbnail.py:
```diff
@@ -0,0 +1,76 @@
+    if source_type == THUMBOR_EXTERNAL_TYPE:
+        url = path
+
+    # Hack to getby weird issue with http in path and nginx.
```
get by (add space)

adnrs96 commented:
This has been updated and is IMO ready for a review.

timabbott commented:
Cool!  I merged the first 2 commits.  For the next commit:
* `requirements/thumbor.txt` probably mostly just needs `thumbor` itself; the rest of those are recursive dependencies that we should be able to get `tools/update-locked-requirements` to add for us...  @HarshitOnGitHub @rht and @hackerkid know this system and can help if you need it, but zulip.readthedocs.io/en/latest/dependencies.html should be able to tell you what you need to know.
* It looks like there's some bugfixes to `create-thumbor-venv` in later commits that could probably be moved earlier.
* The big question for me is whether we actually need a separate virtualenv for thumbor; maybe we just want to add `thumbor` to the main Zulip virtualenv?  Given that it seems the Thumbor code kinda wants to be able to access the main Zulip settings, it seems like we'll want Django in the Thumbor virtualenv anyway.  And I think there's a pretty coherent story we could do that's just "Thumbor uses the same virtualenv but is an independent process", more like how Tornado runs.  What do you think?

The rest are more speculative notes on the later commits.  For the first big commit ("thumbor: Add thumbor microservice on port 9995."), I think we can probably get that merged without a lot of risk with the following approach:
* First, extract an early commit that just creates the `thumbor_key` secret; that's easy to merge and harmless.
* Let's leave off the change adding `include zulip::thumbor` to `zulip::voyager` until a later commit; that way, we can merge things, and test them, without much risk to people installing a normal server off master (since they won't get thumbor at all).  
* This will be super inefficient, because `get-django-setting` requires loading the whole Zulip app to get the values.  
```
+def read_zulip_settings_variable(variable_name):
+    script = os.path.join(ZULIP_PATH, 'scripts', 'get-django-setting')
+    proc = subprocess.Popen([script, variable_name], stdout=subprocess.PIPE)
+    output = proc.stdout.read()
+    try:
+        return eval(output)
+    except Exception:
+        return output.strip()
```

I think we probably want to refactor so we call that at most once.  Better would be 0 times.

timabbott commented:
I did a bit of attempting to put thumbor in the same virtualenv as the rest of Zulip, and got stopped because it apparently isn't compatible with modern Pillow.  So I guess we should keep it in its own virtualenv for now.

rht commented:
> This will benefit mobile devices in terms of performance. 

Sounds like this needs to be merged soon.

adnrs96 commented on zthumbor/loaders/zloader.py:
```diff
@@ -0,0 +1,44 @@
+    result.successful = False
+    return result
+
+@return_future  # type: ignore # This was giving mypy error for missing generic datatype?
```
He mypy was reporting a missing generic type annotation (which I didn't knew how to fix) so for the timing I added a `type: ignore` but I would really like to know what happened here.
To reproduce just remove the type declaration comment and run mypy.

timabbott replied:
@neiljp may be able to help debug this.

neiljp replied:
The simplest fix for this is annotating with `Callback[..., Any]`, but you may know more about the specific parameter and return types based on the context.

timabbott commented:
The requirements stuff looks good, but I got this issue when I tried running it:

```
diff --git a/requirements/thumbor_lock.txt b/requirements/thumbor_lock.txt
index 4384c9c..7155af4 100644
--- a/requirements/thumbor_lock.txt
+++ b/requirements/thumbor_lock.txt
@@ -1,19 +1,16 @@
 #
-# This file is GENERATED.  Don't edit directly.
+# This file is autogenerated by pip-compile
+# To update, run:
 #
-# To update, edit the non-"lock" files in requirements/*.txt, then:
-#
-#    tools/update-locked-requirements
-#
-# For details, see requirements/README.md .
+#    pip-compile --output-file requirements/thumbor_lock.txt requirements/thumbor.txt
 #
 argparse==1.4.0           # via thumbor
-backports-abc==0.5        # via tornado
 backports.ssl-match-hostname==3.5.0.1
 botocore==1.2.0           # via tornado-botocore
 certifi==2016.9.26
 derpconf==0.8.1           # via thumbor
 docutils==0.12            # via botocore
+futures==3.1.1            # via thumbor
 jmespath==0.7.1           # via botocore
 libthumbor==1.3.2         # via thumbor
 pexif==0.15               # via thumbor
```

rht commented:
The `futures` line must be deleted according to the `tools/update-locked-requirements` script.

adnrs96 commented:
ohk yup, so Tim ran the pip-compile directly I guess (after looking at the diff Tim posted).
For the `backports-abc==0.5` being removed from the deps I think this [line of code](https://github.com/tornadoweb/tornado/blob/master/setup.py#L132) might we the answer. Basically I think its not needed in python above 3.5.
@timabbott can you let me know what exact python version you are using?

adnrs96 commented:
hmm, I think `futures==3.1.1` is required since thumbor uses python 2 in a new virtual env.
Maybe we shouldn't use the `tools/update-locked-requirements`?

adnrs96 commented:
So it turns out that we indeed needed the `futures` lib for running thumbor and since it runs in a py2 env the library is not obsolete there. To tackle this this is what I have done:
* Add a new commit which alters the `update-locked-requirements` to have a third argument for the `compile_requirements` function which is the python version we wanna compile for.
* Use this new commit to get the new `thumbor_lock` requirements file.

Some other  stuff that I changed in the recent updates is:
* Remove the `THUMBOR_VENV_DEPS` from being handled by puppet and made the create-thumbor-venv do the installation.

timabbott commented:
I just merged the first 4 commits, after adding `backports-abc` back to the list of things hand-added in thumbor.txt (since `update-locked-requirements` wanted to remove that on `xenial`), thanks @adnrs96!  

I think you can go ahead and squash the commits that add tests into the commits that introduce the code that's tested.

timabbott commented on zerver/lib/thumbnail.py:
```diff
@@ -0,0 +1,84 @@
+    if url.startswith('https:/'):
+        url = url[len('https:/'):]
+    if url.startswith('http:/'):
+        url = url[len('http:/'):]
```
Do you understand this hack @adnrs96?

adnrs96 replied:
consider a situation as follows:
There is a link to an external image say `https://www.google.com/images/srpr/logo4w.png`. When pasted and sent as a message the rendered body of the message will have a block for inline image which will use relative url to refer to this image as: `/thumbnail/https://www.google.com/images/srpr/logo4w.png`. Idea behind the endpoint is any link sent to the endpoint by doing something like `/thumbnail/<link-here>` will get thumbnailed by thumbor.

Thing is when this kind of links hit Nginx, nginx misbehaves.
>What type of link?
>>Links with `https://` or `http://` in there path

So my best guess here is Nginx finds it difficult to parse such links and ends up miss-redirecting such requests. So I placed in this code (or what I like to call hack because it sounds cool!) to handle nginx's misbehaviour.
@timabbott Sorry for the super delayed explanation on this. I thought I already had replied to this question back when you asked.

timabbott replied:
So, first, `https:/` URLs don't really exist, so I'm not sure why you'd need the second half.  But also, this can be done non-hackisly using the `urllib.parse` library -- you just want to get rid of the `netloc` section.

zulipbot-test commented:
Hello @adnrs96, a Zulip maintainer reviewed this pull request over 10 days ago, but you haven't updated your pull request since. Please take a look at the requested changes and update your pull request accordingly.

Thank you for your valuable contributions to Zulip!

<!-- updateWarning -->

timabbott commented:
I merged `thumbor: Add thumbor microservice on port 9995.`, after editing the commit message to be clear it's only running in development.

timabbott commented:
I merged the first commit as ecbb763e795245d4b60b08e2b63dfdddbd618144, after substantial editing for clarity.  

For the next commit, can you explain the `generate_thumbnail_url` logic?

adnrs96 commented:
Hi @timabbott 
Here is an explanation for the logic :)

generate_thumbnail_url is responsible for getting us a URL to which we must redirect our request in order to serve the image. The response image can vary in resolution and this resolution is given as an input for thumbnail link generation If no resolution is given as input a default of 0X0 resolution is assumed which means generate thumbnail with original resolution which will effectively mean
return the original image.
Now thumbor responds with images on receiving requests having a particular path. This request is validated and responded to if it turns out to be valid.

Lets talk about generate_thumbnail_url:
I am gonna list the code flow in points and there function:

* First of all we check if thumbor is even enabled. In case of thumbor service
being disabled we don't go through the pain of a specific url generation and
return a such a url which normally use to access files or links.
So for example if thumbor is disabled and we recieved a thumbnailing request for
a user uploaded file with path `user_uploads/2/7b/NQZTm1wfArMoIBjGi8bZn7CY/_MG_0455.JPG`
we will just return this path and the caller will redirect request to link and image
will be served as it is served right now in the app.
If the request was for a https link the link is returned but if the request was
for a http link a camo link is generated and returned.

* Next we just ignore request that start with 'static/'. We do this because there
are some messages with image links hard coded which are sent to a stream when we
first time install Zulip. To keep those images working we add this exeception.

* Next we decide what source time does our request has. Here sources type doesn't
mean what client was used to make the request but what kind of a link we are looking
at for thumbnailing.
Eg.
Local type link such as user uploaded image files stored in local storage.
Amazon S3 type links with images stored in the amazon cloud storage.
External web site links which are most probably of the form https://some.website/path_to_image

* Next we have a hack for dealing with a problem with Nginx. My explanation for this hack
could be found at https://github.com/zulip/zulip/pull/6012/commits/a68b02f4084fc27dda4e7f884f58b1b920c93a4f#r163644677

* Next we join our url, thumbnail_resolution_requested and the source type into a '_' separated string.
We use this string along with the server generated thumbor key to generate a hash which we kind of use a signature.
This sign_hash is urlencoded along with source_type and sent as query parameters in the url we will redirect the request.

All of the above is now used to put together a link whose format is specified by thumbor. This url is returned to the caller which then redirects the request to the link.

adnrs96 commented on puppet/zulip/files/supervisor/conf.d/thumbor.conf:
```diff
@@ -0,0 +1,12 @@
+[program:zulip-thumbor]
+command=/home/zulip/deployments/current/zulip-thumbor-venv/bin/thumbor --port=9995 --conf=/home/zulip/deployments/current/zthumbor/thumbor.conf
```
I think this is going to be problematic since I have given commands using hard wired path to current directory which might not always be the directory from which app is deployed.

adnrs96 replied:
oops, I think this is just fine. We do symlink to current at deployment.

timabbott commented:
@adnrs96 I did a round of reviewing this; here's a few questions:
* Reading https://github.com/thumbor/thumbor/wiki/Security, it appears we should be setting `SECURITY_KEY` to be our `THUMBOR_KEY`?  Or are we implementing that functionality ourselves by passing `sign_hash = get_sign_hash(raw, settings.THUMBOR_KEY)` ourselves?  I'd like to be sure I understand this part, since security is important.  
* Relatedly, is there a way to not have `/unsafe/` appear in our URLs?  It seems that's the Thumbor default if you're not using their built-in signing, but we definitely want to be doing signature checking...
* That page also mentions the `https_loader` -- is there a reason to prefer the `http_loader`?  Based on http://thumbor.readthedocs.io/en/latest/image_loader.html#pre-packaged-loaders, I think we want default-to-https (but not the strict one, since we do want to thumbnail sites when https is not available, for now).
* Just to confirm my understanding, the actual URLs we're using in the Bugdown markdown processor just have the size hardcoded in the URL; all the other decisions are made in the /thumbnail/ endpoint.  So we can change basically any aspect of the Thumbor settings later (including, e.g., THUMBOR_HOST), other than the sizing decisions.  The downside of this approach is that we effectively require 2 RTTs to fetch an image, due to the redirect.

timabbott commented on scripts/lib/upgrade-zulip-stage-2:
```diff
@@ -123,7 +127,7 @@ if not args.skip_puppet or migrations_needed:
     logging.info("Stopping Zulip...")
     subprocess.check_call(["supervisorctl", "stop", "zulip-workers:*", "zulip-django",
-                           "zulip-tornado", "zulip-senders:*"], preexec_fn=su_to_zulip)
+                           "zulip-tornado", "zulip-senders:*", "zulip-thumbor"], preexec_fn=su_to_zulip)
```
This may cause problems on systems that don't have thumbor installed.  We should probably instead do something similar to what we do with `process-fts-updates` where we make this being in the list conditional.  

Since doing a second supervisord call for this is silly, we should probably just make the list of processes a variable, and include this or not depending on whether the `/etc/supervisord/conf.d` file exists.

(That same treatment doesn't make as much sense for `process-fts-updates`, since that one isn't part of the basically serving code path)

timabbott commented on scripts/restart-server:
```diff
@@ -29,15 +29,15 @@ subprocess.check_call(["./manage.py", "fill_memcached_caches"])
 subprocess.check_call(["supervisorctl", "stop", "zulip-workers:*"])
 logging.info("Stopping server core")
-subprocess.check_call(["supervisorctl", "stop", "zulip-senders:* zulip-django zulip-tornado"])
+subprocess.check_call(["supervisorctl", "stop", "zulip-senders:* zulip-django zulip-tornado zulip-thumbor"])
```
I think the same argument as above applies here.

timabbott commented on zerver/lib/thumbnail.py:
```diff
@@ -0,0 +1,84 @@
+    )
+    if settings.THUMBOR_HOST != '127.0.0.1:9995':
+        thumbnail_url = u'{host}'.format(
+            host=settings.THUMBOR_HOST) + thumbnail_url
```
This is written in odd Python style.  Would also do this with `urllib.parse`.

timabbott commented on zerver/tests/test_thumbnail.py:
```diff
@@ -0,0 +1,207 @@
+        with self.settings(THUMBOR_HOST='test-thumborhost.com'):
+            result = self.client_get("/thumbnail" + uri)
+        self.assertEqual(result.status_code, 302, result)
+        base = 'test-thumborhost.com/thumbor/unsafe/0x0/smart/files/'
```
why not verify the entire URL here?

adnrs96 replied:
Since the entire url will contain a dynamic sign key it is kind of tricky to do the testing without generating the entire hash again here using a copy of logic earlier used.

timabbott commented on zerver/tests/test_thumbnail.py:
```diff
@@ -0,0 +1,207 @@
+            result = self.client_get("/thumbnail" + uri)
+        self.assertEqual(result.status_code, 302, result)
+        base = '/user_uploads/'
+        self.assertEqual(base, result.url[:len(base)])
```
let's verify the entire URL in these tests; I think that'd be more robust to potential future bugs.

timabbott commented on zerver/views/thumbnail.py:
```diff
@@ -0,0 +1,33 @@
+
+def validate_thumbnail_request(user_profile, path):
+    # type: (UserProfile, str) -> bool
+    if path.startswith('user_uploads/'):
```
So one thing that worries me about these checks is that I'm never sure whether there's a leading `/` on `path` without reading a lot of code.  Maybe just deserves a comment here.

Also, we should update the mypy annotations to the Python 3 style.

timabbott commented on zerver/views/thumbnail.py:
```diff
@@ -0,0 +1,33 @@
+        )
+        return redirect(thumbnail_url)
+
+    return HttpResponseForbidden(_("<p>You are not authorized to view this file.</p>"))
```
We should make a note on a list of follow-ups to consider making this a nicer error page.  Definitely don't do anything for now, though; it's not that important.

timabbott commented on zerver/views/thumbnail.py:
```diff
@@ -0,0 +1,33 @@
+    url = urllib.parse.unquote(url)
+    if validate_thumbnail_request(user_profile, url):
+        thumbnail_url = generate_thumbnail_url(
+            url.rsplit('?', 1)[0],
```
This is much cleaner to do with `urllib.parse`  and the `query` parameter.

adnrs96 replied:
actually I realised that the query won't even be there in the url at this point!

timabbott commented on zproject/urls.py:
```diff
@@ -520,6 +520,11 @@
                      {'override_api_url_scheme'})}),
 
+# This endpoint will be used to proxy thumbnail requests to thumbor host in case
+# we are in dev environment or thumbor is not on same server as zulip.
```
Just to confirm, this endpoint is used always, whether they're the same server or not, right?

timabbott commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -216,7 +217,21 @@ def add_a(
+            urllib.parse.quote(url)
+        ))
+    else:
+        img.set("src", url)
```
Are there any call points that are not using `use_thumbnails`?  I kinda wonder if we want to flip this around, so the default is True and we only specify in the cases where we don't want a thumbnail...

adnrs96 commented:
> Just to confirm my understanding, the actual URLs we're using in the Bugdown markdown processor just have the size hardcoded in the URL; all the other decisions are made in the /thumbnail/ endpoint. So we can change basically any aspect of the Thumbor settings later (including, e.g., THUMBOR_HOST), other than the sizing decisions. The downside of this approach is that we effectively require 2 RTTs to fetch an image, due to the redirect.

Yeah, that is correct on all aspects. Making size dynamic would be probably tricky.

adnrs96 commented:
The default `https_loader` provided by thumbor doesn't fallback to http in case of failure to load a image. It just depends on what the original incoming link said it was (http or https), the very information which is lost in our case to handle the nginx issue.
Here is a error log when we try to use default `https_loader`:
```
2018-03-09 07:13:48.979 INFO [django.server] "GET /thumbnail/http%3A//www.ccet.ac.in/images/icons/1200x400_apratim_glimpse.png?size=0x100 HTTP/1.1" 302 0
2018-03-09 07:13:49 thumbor:WARNING ERROR retrieving image https://www.ccet.ac.in/images/icons/1200x400_apratim_glimpse.png: [Errno 104] Connection reset by peer
2018-03-09 07:13:49 tornado.access:ERROR 504 GET /unsafe/0x100/smart/www.ccet.ac.in/images/icons/1200x400_apratim_glimpse.png%3Fsign%3Dtz%252Bs7Yiwhj6%252BXmJMI8lML0Vrjnk%253D%26source_type%3Dexternal (10.0.3.1) 163.12ms
```

Addressed this issue by making a second request for the image manually in case the failure occurred and it was a EXTERNAL_TYPE image (basically append `http` and do the same request again)

adnrs96 commented:
@timabbott I have updated this while addressing your feedback, It contains some commits which are meant to (maybe) make reviewing process easy. Those commits begin with a squash and I can squash them to the correct commits when you are ready.
Apart from that I addressed the signing process issue. For that I ended up using the internal signing option of thumbor. I did continue to include the source type in the signing process by appending it the image path. This way we kinda make it work for us in all ways.
Also added a commit after which I think S3 storage option should work. I couldn't really test it in action.
But I ensured the;
* path to thumbor s3 loader is correct,
* S3 key and pass are loaded up to thumbor in the config file (already merged)
So I think it should work.

(It has been really great working on this issue :))

sampritipanda commented on zthumbor/loaders/helpers.py:
```diff
@@ -44,43 +41,13 @@ def get_secret(key):
+def separate_url_and_source_type(url):
+    # type: (Text) -> Tuple[Text, Text]
+    THUMBNAIL_URL_PATT = re.compile('^(?P<actual_url>.+)/source_type/(?P<source_type>.+)')
+    matches = THUMBNAIL_URL_PATT.match(url)
```
typo here.

timabbott commented:
Merged!  Huge thanks to @sampritipanda and @adnrs96 for your work on this!  And thanks to @TigorC for the original work and @joshland for first getting this fully working at PyCon 2017.

There's a few follow-up things we need to do:
* Consider renaming "original" to "full" as the name for the "full" size.
* Enable it on chat.zulip.org and test it out (and implement whatever mobile needs)
* Work on the tooling for testing zthumbor/; we should at least get the linter running against that if we can.

This is pretty exciting, since it resolves one of the longest-open issues in Zulip (#432)! :)
